synapse/handlers/message.py

```diff
@@ -906,7 +906,18 @@ def persist_and_notify_client_event(
             directory_handler = self.hs.get_handlers().directory_handler
             if room_alias_str and room_alias_str != original_alias:
                 room_alias = RoomAlias.from_string(room_alias_str)
-                mapping = yield directory_handler.get_association(room_alias)
+                try:
+                    mapping = yield directory_handler.get_association(room_alias)
+                except SynapseError as e:
+                    # Turn M_NOT_FOUND errors into M_BAD_ALIAS errors.
```

Comment:
have you checked that this codepath is reached when it's a remote alias? I wouldn't be surprised if we fail to turn the HTTP 404 into a SynapseError.

Reply:
I have not checked that. Will look into it!

Reply:
I believe this is properly handled:

https://github.com/matrix-org/synapse/blob/229eb81498b0fe1da81e9b5b333a0285acde9446/synapse/handlers/directory.py#L240-L264

That code seems to handle the 404, never setting a `room_id`, which then raises the `SynapseError`.